src/sign/generic.rs

```diff
@@ -0,0 +1,512 @@
+                .map_err(|_| DnsFormatError::Misformatted)?;
+
+            *field = Some(buffer.into());
+            data = rest;
+        }
+
+        for field in [&n, &e, &d, &p, &q, &d_p, &d_q, &q_i] {
+            if field.is_none() {
+                // A field was missing.
+                return Err(DnsFormatError::Misformatted);
+            }
+        }
+
+        Ok(Self {
+            n: n.unwrap(),
+            e: e.unwrap(),
+            d: d.unwrap(),
+            p: p.unwrap(),
+            q: q.unwrap(),
+            d_p: d_p.unwrap(),
+            d_q: d_q.unwrap(),
+            q_i: q_i.unwrap(),
+        })
+    }
+}
+
+impl<B: AsRef<[u8]> + AsMut<[u8]>> Drop for RsaSecretKey<B> {
+    fn drop(&mut self) {
+        // Zero the bytes for each field.
+        self.n.as_mut().fill(0u8);
+        self.e.as_mut().fill(0u8);
+        self.d.as_mut().fill(0u8);
+        self.p.as_mut().fill(0u8);
+        self.q.as_mut().fill(0u8);
+        self.d_p.as_mut().fill(0u8);
+        self.d_q.as_mut().fill(0u8);
+        self.q_i.as_mut().fill(0u8);
+    }
+}
+
+/// A generic public key.
+pub enum PublicKey<B: AsRef<[u8]>> {
+    /// An RSA/SHA-1 public key.
+    RsaSha1(RsaPublicKey<B>),
+
+    // TODO: RSA/SHA-1 with NSEC3/SHA-1?
+    /// An RSA/SHA-256 public key.
+    RsaSha256(RsaPublicKey<B>),
+
+    /// An RSA/SHA-512 public key.
+    RsaSha512(RsaPublicKey<B>),
+
+    /// An ECDSA P-256/SHA-256 public key.
+    ///
+    /// The public key is stored in uncompressed format:
+    ///
+    /// - A single byte containing the value 0x04.
+    /// - The encoding of the `x` coordinate (32 bytes).
+    /// - The encoding of the `y` coordinate (32 bytes).
+    EcdsaP256Sha256([u8; 65]),
+
+    /// An ECDSA P-384/SHA-384 public key.
+    ///
+    /// The public key is stored in uncompressed format:
+    ///
+    /// - A single byte containing the value 0x04.
+    /// - The encoding of the `x` coordinate (48 bytes).
+    /// - The encoding of the `y` coordinate (48 bytes).
+    EcdsaP384Sha384([u8; 97]),
+
+    /// An Ed25519 public key.
+    ///
+    /// The public key is a 32-byte encoding of the public point.
+    Ed25519([u8; 32]),
+
+    /// An Ed448 public key.
+    ///
+    /// The public key is a 57-byte encoding of the public point.
+    Ed448([u8; 57]),
+}
+
+impl<B: AsRef<[u8]>> PublicKey<B> {
+    /// The algorithm used by this key.
+    pub fn algorithm(&self) -> SecAlg {
+        match self {
+            Self::RsaSha1(_) => SecAlg::RSASHA1,
+            Self::RsaSha256(_) => SecAlg::RSASHA256,
+            Self::RsaSha512(_) => SecAlg::RSASHA512,
+            Self::EcdsaP256Sha256(_) => SecAlg::ECDSAP256SHA256,
+            Self::EcdsaP384Sha384(_) => SecAlg::ECDSAP384SHA384,
+            Self::Ed25519(_) => SecAlg::ED25519,
+            Self::Ed448(_) => SecAlg::ED448,
+        }
+    }
+
+    /// Construct a DNSKEY record with the given flags.
+    pub fn into_dns<Octs>(self, flags: u16) -> Dnskey<Octs>
```

Comment:
I want to avoid the same public key being transformed into DNSKEY records with different flags values (since that is almost certainly incorrect).  I'm going to address this a bit by storing that metadata in the primary public key type directly in an overhaul of this PR.